graph-builder/src/registry.rs

```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_path: &Option<PathBuf>,
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(|| -> Result<Box<dyn std::io::Read>, Error> {
+            match credentials_path {
+                Some(path) => match File::open(path) {
+                    Ok(file) => Ok(Box::new(file)),
+                    Err(e) => Err(e.into()),
+                },
+                None => Ok(Box::new(std::io::empty())),
+            }
+        }()?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let authenticated_client = tcore
+        .run(authenticate_client(
+            &mut client,
+            format!("repository:{}:pull", &repo),
+        ))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let tags = match tcore.run(
+        authenticated_client
+            // According to https://docs.docker.com/registry/spec/api/#listing-image-tags
+            // the tags should be ordered lexically but they aren't
+            .get_tags(&repo, Some(20))
+            .map_err(|e| format_err!("{}", e))
+            .collect(),
+    ) {
+        Ok(mut tags) => {
+            if tags.len() == 0 {
+                bail!("{}/{} has no tags", registry, repo);
+            }
+            tags.sort();
+            Ok(tags)
+        }
+        Err(e) => Err(e),
+    }?;
+
+    let releases = futures::stream::iter_ok(tags.to_owned().into_iter())
+        .and_then(|tag| {
+            trace!("processing: {}:{}", &repo, &tag);
+            authenticated_client
+                .has_manifest(&repo, &tag, None)
+                .join(authenticated_client.get_manifest(&repo, &tag))
+                .map(|(manifest_kind, manifest)| {
+                    get_layers_for_manifest_with_kind(manifest_kind, manifest, tag)
+                })
+                .map_err(|e| format_err!("{}", e))
+        })
+        .and_then(|tag_layer_digests| match tag_layer_digests {
+            Ok((tag, layer_digests)) => {
+                if layer_digests.len() == 0 {
+                    bail!("layer_digests list empty")
+                };
+                Ok((tag, layer_digests))
+            }
+            Err(e) => bail!("{}", e),
         })
+        .and_then(|(tag, layer_digests)| {
+            find_first_release_in_layers(
+                authenticated_client.clone(),
+                layer_digests,
+                registry_host.into(),
+                repo.into(),
+                tag,
+            )
+        })
+        .collect();
+
+    let releases = tcore.run(releases)?;
+    Ok(releases)
+}
+
+fn get_layers_for_manifest_with_kind(
+    manifest_kind: Option<dkregistry::mediatypes::MediaTypes>,
+    manifest: Vec<u8>,
+    tag: String,
```

Comment:
Can we remove this argument since it isn't needed?

Reply:
It's passed through to the `Release { ... }` creation.

Reply:
I don't follow. I was able to remove this in my test branch and create the tuple up at the call site. That seems like the more appropriate place to inject the tag.

Reply:
That works too